### ISSUE TITLE
conda update downgrades packages

### ISSUE THREAD

**Issue by jakirkham** [conda/conda#2263]
If I have something installed that tracks features, due to the `mkl` situation I have found myself needing to track `nomkl` and `openblas` as features. Maybe there is a better way, but this is currently working for me. I sometimes end up in a situation where features are removed when something depends on something tracking features. Though this is resolvable by updating, it is preferable if it doesn't happen. Here is an example.

```
$ conda install -y seaborn
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: ..........

Package plan for installation in environment /opt/conda2:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openblas-0.2.15            |              100        14.9 MB  nanshe
    numpy-1.10.4               |   py27_nomkl_106         6.6 MB  nanshe
    scipy-0.17.0               |   py27_nomkl_102        31.1 MB  nanshe
    pandas-0.18.0              |      np110py27_0        12.0 MB  defaults
    scikit-learn-0.17.1        |   py27_nomkl_102         9.2 MB  nanshe
    spams-v2.5                 |    py27_nomkl_13         4.4 MB  nanshe
    seaborn-0.7.0              |           py27_0         264 KB  defaults
    ------------------------------------------------------------
                                           Total:        78.3 MB

The following NEW packages will be INSTALLED:

    pandas:       0.18.0-np110py27_0    defaults
    seaborn:      0.7.0-py27_0          defaults

The following packages will be DOWNGRADED:

    numpy:        1.10.4-py27_nomkl_openblas_107 nanshe [nomkl openblas] --> 1.10.4-py27_nomkl_106 nanshe   [nomkl]
    openblas:     0.2.15-nomkl_103               nanshe [nomkl         ] --> 0.2.15-100            nanshe  
    scikit-learn: 0.17.1-py27_nomkl_openblas_103 nanshe [nomkl openblas] --> 0.17.1-py27_nomkl_102 nanshe   [nomkl]
    scipy:        0.17.0-py27_nomkl_openblas_103 nanshe [nomkl openblas] --> 0.17.0-py27_nomkl_102 nanshe   [nomkl]
    spams:        v2.5-py27_nomkl_openblas_14    nanshe [nomkl openblas] --> v2.5-py27_nomkl_13    nanshe   [nomkl]

Fetching packages ...
openblas-0.2.1 100% |#############################################################################################################| Time: 0:00:01   7.86 MB/s
numpy-1.10.4-p 100% |#############################################################################################################| Time: 0:00:01   6.00 MB/s
scipy-0.17.0-p 100% |#############################################################################################################| Time: 0:00:03  10.71 MB/s
pandas-0.18.0- 100% |#############################################################################################################| Time: 0:00:01   6.78 MB/s
scikit-learn-0 100% |#############################################################################################################| Time: 0:00:01   8.83 MB/s
spams-v2.5-py2 100% |#############################################################################################################| Time: 0:00:01   4.44 MB/s
seaborn-0.7.0- 100% |#############################################################################################################| Time: 0:00:00   1.55 MB/s
Extracting packages ...
[      COMPLETE      ]|################################################################################################################################| 100%
Unlinking packages ...
[      COMPLETE      ]|################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|################################################################################################################################| 100%
$ conda update --all -y
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: ..........

Package plan for installation in environment /opt/conda2:

The following packages will be UPDATED:

    numpy:        1.10.4-py27_nomkl_106 nanshe [nomkl] --> 1.10.4-py27_nomkl_openblas_107 nanshe [nomkl openblas]
    openblas:     0.2.15-100            nanshe --> 0.2.15-nomkl_103               nanshe [nomkl         ]
    scikit-learn: 0.17.1-py27_nomkl_102 nanshe [nomkl] --> 0.17.1-py27_nomkl_openblas_103 nanshe [nomkl openblas]
    scipy:        0.17.0-py27_nomkl_102 nanshe [nomkl] --> 0.17.0-py27_nomkl_openblas_103 nanshe [nomkl openblas]
    spams:        v2.5-py27_nomkl_13    nanshe [nomkl] --> v2.5-py27_nomkl_openblas_14    nanshe [nomkl openblas]

Unlinking packages ...
[      COMPLETE      ]|################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|################################################################################################################################| 100%
```

**Comment by mcg1969** [comment]
I'm going to have to think about this one. I'm not sure we can properly handle a case like this.

**Comment by mcg1969** [comment]
In my view, we need to minimize or eliminate the situations where a `track_features` entry is in a package _that other packages depend on_. They need to be limited to true metapackages.

**Comment by jakirkham** [comment]
> In my view, we need to minimize or eliminate the situations where a `track_features` entry is in a package *that other packages depend on*. They need to be limited to true metapackages.

Unfortunately, this doesn't work for me.

The main reason I have had to do this is because the Accelerate Framework on Mac is not fork safe. However, I need to be able to fork in Python code and need to make sure I have a BLAS that does not suffer from these issues. If I don't do this, end users see sporadic segmentation faults, which is not ok. So, I have used OpenBLAS which is the only reasonably stable, performant, open source BLAS that I am aware of that is fork safe. My Python code needs to indicate that this is a required dependency somehow so I have used this strategy.

If there is a better solution, I am open to trying something different.

**Comment by mcg1969** [comment]
I'm afraid I don't understand why that contradicts my claim. What problems are _created_ if we move all track_feature values out into standalone metapackages?

Let's look at this another way: suppose track_features were not added by packages at all, perhaps they were set on a per-environment basis. No packages could have `track_features`. Could you not find a solution that works in that case?

**Comment by mcg1969** [comment]
As for your example above, can you give me a `conda create` command to set up the initial environment to reproduce the issue? I'm certainly happy to study it.

**Comment by jakirkham** [comment]
Maybe I am misunderstanding what is being said here. The problem is in my Python package I need to be able to say the only permitted BLAS is OpenBLAS. So it needs to *depend* on OpenBLAS somehow.

**Comment by mcg1969** [comment]
Well sure, but that's just a dependency. That doesn't track the feature.

**Comment by jakirkham** [comment]
However, if I now want to have NumPy and SciPy link to it instead of Accelerate or MKL...

**Comment by mcg1969** [comment]
What package has the openblas `track_feature`? I can't find it in the default repo.

**Comment by jakirkham** [comment]
Mine do.

**Comment by mcg1969** [comment]
OK. So which packages did you put it in?

**Comment by mcg1969** [comment]
Ah, I've found the nanshe repodata, I can study it.

**Comment by jakirkham** [comment]
> As for your example above, can you give me a `conda create` command to set up the initial environment to reproduce the issue? I'm certainly happy to study it.

    conda create -c nanshe -n <ENV> -y nanshe && conda update -c nanshe <ENV> --all

**Comment by jakirkham** [comment]
Also, I have a docker container if you would prefer that. ( https://hub.docker.com/r/nanshe/nanshe/ )

**Comment by mcg1969** [comment]
It would not surprise me if part of the problem is that there are channel collisions between _our_ openblas package (which has no track_feature) and yours (which does).

So my recommendation is to remove that track_feature from your packages, and place it instead into a separate metapackage. To start, try installing that package in parallel with `nanshe` and see what happens. Then consider making that package a dependency of the `nanshe` package to see if that does it.

**Comment by mcg1969** [comment]
OK. I've run out of time today and I won't have an opportunity to look at this through the remainder of the week. I _do_ however want to understand what's going on here very well. So please don't close this until then, unless you happen to find a solution you're very happy with.

**Comment by jakirkham** [comment]
That's fine. I appreciate you taking a look at it.

**Comment by jakirkham** [comment]
FWIW, I didn't think I saw this in earlier versions of `conda`. I am not sure exactly when this behavior began. As I am a little hazy on when the changes to the solver entered `conda` in terms of versions I tried using the latest of `conda 3.18.x`. It turns out this doesn't occur then. See below for an example.

```
$ conda --version
conda 3.18.9
$ conda install seaborn
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: ........................................
Package plan for installation in environment /opt/conda2/envs/nanshenv:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    pandas-0.18.0              |      np110py27_0        12.0 MB  defaults
    seaborn-0.7.0              |           py27_0         264 KB  defaults
    ------------------------------------------------------------
                                           Total:        12.3 MB

The following NEW packages will be INSTALLED:

    pandas:  0.18.0-np110py27_0 defaults
    seaborn: 0.7.0-py27_0       defaults

Proceed ([y]/n)? ^C
Operation aborted.  Exiting.
$ source activate root
discarding /opt/conda2/envs/nanshenv/bin from PATH
prepending /opt/conda2/bin to PATH
$ conda update conda
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: .................
Package plan for installation in environment /opt/conda2:

The following packages will be UPDATED:

    conda:  3.18.9-py27_0 defaults --> 4.0.4-py27_0 defaults
    python: 2.7.10-2      defaults --> 2.7.11-0     defaults

Proceed ([y]/n)? 

Unlinking packages ...
[      COMPLETE      ]|################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|################################################################################################################################| 100%
$ source activate nanshenv
discarding /opt/conda2/bin from PATH
prepending /opt/conda2/envs/nanshenv/bin to PATH
$ conda install seaborn
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: ..........

Package plan for installation in environment /opt/conda2/envs/nanshenv:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openblas-0.2.15            |              100        14.9 MB  nanshe
    numpy-1.10.4               |   py27_nomkl_106         6.6 MB  nanshe
    scipy-0.17.0               |   py27_nomkl_102        31.1 MB  nanshe
    pandas-0.18.0              |      np110py27_0        12.0 MB  defaults
    scikit-learn-0.17.1        |   py27_nomkl_102         9.2 MB  nanshe
    spams-v2.5                 |    py27_nomkl_13         4.4 MB  nanshe
    seaborn-0.7.0              |           py27_0         264 KB  defaults
    ------------------------------------------------------------
                                           Total:        78.3 MB

The following NEW packages will be INSTALLED:

    pandas:       0.18.0-np110py27_0    defaults
    seaborn:      0.7.0-py27_0          defaults

The following packages will be DOWNGRADED:

    numpy:        1.10.4-py27_nomkl_openblas_107 nanshe [nomkl openblas] --> 1.10.4-py27_nomkl_106 nanshe   [nomkl]
    openblas:     0.2.15-nomkl_103               nanshe [nomkl         ] --> 0.2.15-100            nanshe  
    scikit-learn: 0.17.1-py27_nomkl_openblas_103 nanshe [nomkl openblas] --> 0.17.1-py27_nomkl_102 nanshe   [nomkl]
    scipy:        0.17.0-py27_nomkl_openblas_103 nanshe [nomkl openblas] --> 0.17.0-py27_nomkl_102 nanshe   [nomkl]
    spams:        v2.5-py27_nomkl_openblas_14    nanshe [nomkl openblas] --> v2.5-py27_nomkl_13    nanshe   [nomkl]

Proceed ([y]/n)? ^C
Operation aborted.  Exiting.
```

**Comment by mcg1969** [comment]
```
conda create -n nanshe -c nanshe nanshe
conda install -n nanshe -c nanshe seaborn
```
I'm not seeing any downgrades here (using my patched version of conda):
```

The following NEW packages will be INSTALLED:

    pandas:  0.18.0-np110py27_0
    seaborn: 0.7.0-py27_0      
```
I do see it with `conda update -n nanshe -c nanshe -all` though.

**Comment by jakirkham** [comment]
Yeah, I only see it with the update, as well. There is something funny about it tracking two features instead of one.

**Comment by mcg1969** [comment]
Actually, what I'm finding is that the patched version doesn't do what you want it to do with 
```
conda create -n nanshe -c nanshe nanshe
```
It's pulling the `nomkl` feature, not the `openblas` feature. In that case, I'm inclined to believe it's *doing the right thing*. One of conda's objectives is to minimize the number of active features required to satisfy the constraints. That's rather important for handling the features used with the Windows `python` package. 

Apparently, `nanshe 0.1.0a50` isn't doing enough to enforce the `openblas` track feature.

I think that if `nanshe` itself was the one that held the `openblas` track_feature instead of `openblas` itself, it might work.

**Comment by jakirkham** [comment]
Sorry, which patch? This one ( https://github.com/conda/conda/pull/2260 )?

I think that patch is ok for solving ( https://github.com/conda/conda/issues/2259 ). Though it is has no effect on this issue, so I remain in favor of that patch and simply think this is an orthogonal issue.

> One of conda's objectives is to minimize the number of active features required to satisfy the constraints.

So, I'm not clear on when we came to the conclusion that it was right to minimize features.

In any event, I don't think this is the desired behavior. One of the features that was installed intentionally is getting removed.

> I think that if `nanshe` itself was the one that held the `openblas` `track_feature` instead of `openblas` itself, it might work.

I don't follow why `nanshe` should have `track_feature: - openblas`. My understanding of `track_feature` is this is added to the package that is installed to enable the feature and thus select variants of packages for the environment that have the `feature` selected. So, it seems that `openblas` should have the `track_feature`.

If you were meaning that `nanshe` should have the `feature` `openblas` enabled, then that I can follow and you may be right. So, I went ahead and tried this and added [`nomkl` and `openblas`]( https://github.com/jakirkham/nanshe/commit/00f66baa536b420a9fb60276e04c151ad42358db ). However, I run into a [similar problem]( https://travis-ci.org/jakirkham/nanshe/jobs/116279405#L1062 ).

**Comment by mcg1969** [comment]
> So, I'm not clear on when we came to the conclusion that it was right to minimize features. Maybe you can clarify your reasoning here or maybe I have simply missed it on one of the many threads. :)

The reason we need to minimize the number of features is that some packages have multiple `track_features` possibilities, and we need to make sure that no more than one of those is selected. This is particularly critical for Windows, where Python itself relies on features to control the MSVC runtime selection. It is very unlikely this is going to change, therefore.

The problem right now is that while _you_ are adding the `openblas` track_feature to the `openblas` package, the default channel is not. During a `conda create ... nanshe`, `nanshe` has a choice: install a version with the feature, and install a version without it. It is going to prefer to install the feature-less version of the package.

Now, it is true that your OpenBLAS has a later version number than default; that's true. However, _you didn't ask to maximize the version_ of `openblas`. The order of preference here is this:

1. Maximize the versions of _requested_ packages (nanshe)
2. Minimize the version count (mkl is selected, but not openblas)
3. Maximize the number of packages with the selected features
4. Maximize the version numbers of the packages

On the other hand, when you do `conda update --all`, you _are_ asking to maximize the version of OpenBLAS; and when you do that, you get the `openblas` track_feature. That's why `conda update --all` is producing a different solution than the initial `conda create`.

The solution here, in my view, is clear: remove the `openblas` track_feature from your builds of OpenBLAS and put it somewhere else. I'm recommending `nanshe` itself, but you don't have to do that if you don't want to. But it needs to be somewhere where you can guarantee it gets pulled in by `nanshe`.

Another solution is to hard-code the OpenBLAS version into your `nanshe` recipe, so that it makes sure to pull in your build of OpenBLAS. But that is fragile, in that it might break the minute the default channel updates OpenBLAS.

**Comment by mcg1969** [comment]
Once we have proper channel collision handling---so that your versions of OpenBLAS can _completely_ "shadow" the default channnel---this problem will go away, too. But I don't have a timeline for this.

**Comment by jakirkham** [comment]
Ok, thanks for the above explanation, how this is working is becoming clearer to me.

I have given this more thought and would like to start recodifying things and would appreciate your input as I go forward. As the movement is towards putting in conda forge, I would like to starting adding recipes under some newer mutually agreeable and will appreciate your feedback throughout the process anywhere you are able to give it.

One of the main things that I have been thinking about is that features provides IMO is an ability to select API implementations. As in this case, BLAS/LAPACK implementations. Given this I am thinking that using your metapackage solution is the right way to go. Currently am leaning towards a naming scheme like `blas_*` for BLAS metapackages (so `blas_openblas` in this case). Possibly will add a LAPACK metapackage feature as well (i.e. `lapack_*`). It would be nice to subsume `mkl`/`nomkl` in this framework, which could be doable with meta packages for them, but may need some more thought.

In the future, it may be possible to capture this logic in the `meta.yaml`/solver by explicitly specifying certain classes of API like a BLAS, LAPACK, readline, etc. and then only allowing one implementation of that API. At present, this naming scheme combined with minimization of total features should still give the right behavior.

What are your thoughts on this proposal?

**Comment by mcg1969** [comment]
This is actually very similar to the MSVC issue on Windows. I do think features, especially with your `blas_` prefix convention, could be used to accomplish the desired outcome from a practical perspective.

But to do this _right_, I think, requires improvements to conda and conda-build, along the lines of what you're saying in the last paragraph. What we've discussed is having a new metadata field called `provides`. MKL and OpenBlas would have something like `provides="blas"`, the various C runtimes would have `provides="crt"`, etc. This would be used to provide the mutual exclusivity that you really need in a case like this.

After all, in theory, you can install multiple MSVC runtimes into the same environment, multiple BLAS versions into the same environment. They're different packages, after all. But you _do_ want to make sure that only one is _active_ in any given package set. Here's the tough case: have Python + NumPy + SciPy depend on OpenBlas and MSVCRT14, and R depend on MKL and MSVCRT15, all in the same `conda` environment...

**Comment by mcg1969** [comment]
@jakirkham, do you think we should close this one?

**Comment by jakirkham** [comment]
Let me try and work on the new BLAS recipe convention and then we can discuss here. Once we are happy with the proposal, I plan on moving that directly into conda-forge. In other words, right now I feel this is not so much your (or `conda`'s) issue as mine and it just helps me to have a place to track this stuff until we figure out a recipe convention that is acceptable.

Though the exclusion point discussed above might be something we need to track somewhere. We can move that out into another issue if you wish. I will leave that up to you.

**Comment by kalefranz** [comment]
Believe to be resolved with conda `4.0.5` and https://github.com/conda/conda/pull/2260.  Closing.
